# Error Handling

- Error Handling adalah sebuah proses untuk menangani error yang terjadi pada program kita. Error handling sangat penting untuk dilakukan karena jika kita tidak menangani error yang terjadi, program kita akan berhenti dan tidak akan berjalan lagi.

## 1. try ... catch

- `try ... catch` adalah sebuah statement yang digunakan untuk menangani error yang terjadi pada program kita. `try ... catch` memiliki 2 bagian, yaitu `try` dan `catch`. `try` digunakan untuk menuliskan kode yang akan kita coba jalankan. `catch` digunakan untuk menuliskan kode yang akan kita jalankan jika terjadi error pada kode yang ada di dalam `try`.

  ```js
  try {
    // code yang akan kita coba jalankan
  } catch (error) {
    // kode yang akan kita jalankan jika terjadi error
  }
  ```

- Error Object adalah sebuah object yang digunakan untuk menyimpan informasi error yang terjadi pada program kita. Error Object memiliki 2 properti, yaitu `name` dan `message`. `name` digunakan untuk menyimpan nama error yang terjadi. `message` digunakan untuk menyimpan pesan error yang terjadi.

  ```js
  try {
    // code yang akan kita coba jalankan
  } catch (error) {
    console.log(error.name); // ReferenceError
    console.log(error.message); // myVar is not defined
  }
  ```

```javascript
// try ... catch example
try {
  myVar;
} catch (error) {
  console.log(error.name); // ReferenceError
  console.log(error.message); // myVar is not defined
}

// studi kasus 2
try {
  const myVar = 1;
  myVar = 2;
} catch (error) {
  console.log(error.name); // TypeError
  console.log(error.message); // Assignment to constant variable.
}
```

## 2. Try ... catch ... finally

- `try ... catch ... finally` adalah sebuah statement yang digunakan untuk menangani error yang terjadi pada program kita. `try ... catch ... finally` memiliki 3 bagian, yaitu `try`, `catch`, dan `finally`. `try` digunakan untuk menuliskan kode yang akan kita coba jalankan. `catch` digunakan untuk menuliskan kode yang akan kita jalankan jika terjadi error pada kode yang ada di dalam `try`. `finally` digunakan untuk menuliskan kode yang akan kita jalankan setelah kode yang ada di dalam `try` atau `catch` selesai dijalankan.

  ```js
  try {
    // code yang akan kita coba jalankan
  } catch (error) {
    // kode yang akan kita jalankan jika terjadi error
  } finally {
    // kode yang akan kita jalankan setelah kode di dalam try atau catch selesai dijalankan
  }
  ```

```javascript
// try ... catch ... finally example
try {
  myVar;
} catch (error) {
  console.log(error.name); // ReferenceError
  console.log(error.message); // myVar is not defined
} finally {
  console.log("finally");
}

// output
// ReferenceError
// myVar is not defined
// finally
```

## 3. Throw

- `throw` adalah sebuah statement yang digunakan untuk membuat error yang kita inginkan. `throw` digunakan untuk membuat error yang kita inginkan dan kita bisa menentukan nama error dan pesan error yang kita inginkan.

  ```js
  throw new Error("message");
  ```

  ```javascript
  // throw example
  try {
    throw new Error("my error");
  } catch (error) {
    console.log(error.name); // Error
    console.log(error.message); // my error
  }

  // example 2
  function add(x, y) {
    if (typeof x !== "number") {
      throw "The first argument must be a number";
    }
    if (typeof y !== "number") {
      throw "The second argument must be a number";
    }

    return x + y;
  }

  const result = add("a", 10);

  try {
    console.log(result);
  } catch (error) {
    console.log(error);
  }

  // output :
  // The first argument must be a number
  ```

## 4. Optional Catch Binding

- Optional Catch Binding adalah sebuah fitur yang digunakan untuk membuat error object menjadi optional. Optional Catch Binding digunakan untuk membuat error object menjadi optional. Jika kita tidak ingin membuat error object, kita bisa menghapus parameter error pada `catch`.

  ```js
  try {
    // code yang akan kita coba jalankan
  } catch {
    // kode yang akan kita jalankan jika terjadi error
  }
  ```

  ```javascript
  // optional catch binding example
  try {
    myVar;
  } catch {
    console.log("error");
  }

  // output
  // error
  ```
